iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,327 @@
+                    Ok(data) => data,
+                    Err(e) => {
+                        warn!("Error encrypting message! {}", e);
+                        self.state = State::Error;
+                        return;
+                    }
+                }
+            }
+        };
+        let mut write_half = self.write.as_mut();
+        if let Err(e) = send_message(&mut write_half.unwrap(), data.as_slice()).await {
+            warn!("Error sending message to peer! {}", e);
+            self.state = State::Error;
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl ContextHandler<Connect> for Peer {
+    type Result = ();
+
+    async fn handle(&mut self, ctx: &mut Context<Self>, msg: Connect) {
+        if let Some(addr) = msg.addr {
+            let stream = TcpStream::connect(addr.clone()).await;
+            match stream {
+                Ok(stream) => {
+                    let (read, write) = stream.into_split();
+                    self.read = Some(read);
+                    self.write = Some(write);
+                    self.state = State::Connecting;
+                    let mut write_half = self.write.as_mut();
+                    send_client_hello(&mut write_half.unwrap(), self.public_key.as_bytes()).await;
+                    // Wait for server hello
+                    let mut read_half = self.read.as_mut().unwrap();
+                    read_half.as_ref().readable().await;
+                    if let Ok(message) = read_server_hello(&mut read_half).await {
+                        match message {
+                            Message::ServerHello { key, nonce } => {
+                                let public_key: PublicKey = PublicKey::from(key);
+                                let shared = self.secret_key.diffie_hellman(&public_key);
+                                let chacha = Chacha::new(shared.as_bytes(), &nonce);
+                                self.cipher = Some(chacha);
+                                self.state = State::Ready;
+                                ctx.notify_with(self.stream());
+                            },
+                            _ => {
+                                warn!("Error reading server hello!");
+                                self.state = State::Error;
+                            }
+                        }
+                    }
+                }
+                Err(e) => {
+                    warn!("Could not connect to peer on {}! {}", addr, e);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum State {
+    Connecting,
+    Connected,
+    ClientHelloSent,
+    ServerHelloSent,
+    Ready,
+    Error
+}
+
+pub async fn readable<T: AsRef<TcpStream>>(stream: T) -> io::Result<()> {
+    stream.as_ref().readable().await
+}
+
+pub async fn read_client_hello(stream: &mut OwnedReadHalf) -> Result<Message, Error> {
+    let size = stream.read_u8().await? as usize;
+    if size < 256 {
+        // Reading garbage
```

Comment:
You should make type for garbage. Also you could add doccoment why do we need it

Reply:
Also please move raw values like `256` into named constants.

Reply:
Garbage is not stored anywhere, so it doesn't need a type.
Just for some DPI to be upset because it is harder to invent some pattern.

Reply:
Get it. But still it stored and send via network. So it can be implemented as some type.

Reply:
It is not stored.

Reply:
You literally read it into buffer and not skipping it. This type can be used as documentation. It is also would be good to cover this decision about DPI in docs of this type

Reply:
Also reading garbage kinda repeats

Reply:
> 
> 
> You literally read it into buffer and not skipping it. 

But there is no way to skip it otherwise :(